wave_lang/kernel/wave/wave.py

```diff
@@ -698,29 +702,31 @@ def compile_to_mlir(
         # Only emit MLIR if we don't have a module yet.
         if not module_op:
             emitter = WaveEmitter(
-                dispatch_entrypoint, trace, self.constraints, options, self.grid_type
+                dispatch_entrypoint,
+                trace,
+                self.constraints,
+                options,
+                self.grid_type.dims,
             )
-            try:
+            with mb.module_op.context, Location.unknown():
+                module_op = Module.create()
+
+            with InsertionPoint(module_op.body), Location.unknown():
                 emitter.emit(trace.get_root_graph())
-            except:
-                logger.info("Error in emitter")
-                asm = mb.module_op.get_asm()
-                logger.info(asm)
-                raise
-            emitter.finish()
+
         # Otherwise, we need to iree-fy the existing module (that supposedly has
         # upstream MLIR ops only) in order for it to be executable in the wave
         # pipeline.
         # `dispatch_entrypoint` already has most of the setup, we'll just need
         # to move the ops from existing module to inside `dispatch_entrypoint`.
         # Also we'll need to update the uses of the memref arguments (from the
         # existing module) to be compatible with the new stream.binding arguments.
-        else:
-            assert not any(
-                isinstance(op, stream_d.ExecutableOp)
-                for op in module_op.operation.regions[0].blocks[0]
-            ), "expected overriding module to contain only upstream MLIR ops"
-            _rewrite_module_for_iree_stream_abi(module_op, dispatch_entrypoint, exe)
+
+        assert not any(
+            isinstance(op, stream_d.ExecutableOp)
+            for op in module_op.operation.regions[0].blocks[0]
+        ), "expected overriding module to contain only upstream MLIR ops"
```

Comment:
This now applies to the generated module as well, not only the overriding module

Reply:
yes, that's intentional as `emitter.emit` now will emit fully upstream func and we need to wrap it in `binding_subspan`.